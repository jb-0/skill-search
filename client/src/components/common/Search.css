.search-container {
  width: 400px;
  margin: auto;
}

#searchInputText {
  width: 80%;
  height: 40px;
  padding: 0 20px;
  color: #023047;
  font-size: 1.5rem;
  border-radius: 25px;
  outline-width: 0;
  border: solid;
  border-color: #023047;
}

.suggested-term-container {
  width: 80%;
  background-color: white;
  margin: auto;
  padding: 10px 0;
}

.suggested-term-container p {
  margin: 0;
}

.suggested-term-container:hover {
  color: #fb8500;
  background-color: #8ecae6;
}

.search-term-container {
  background-color: white;
  display:inline-block;
  background-color: #fb8500;
  margin: 5px;
  padding: 0 10px;
  border-radius: 15px;
}
