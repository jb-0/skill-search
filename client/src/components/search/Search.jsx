import React, { useContext } from 'react';
import SearchBox from './SearchBox';
import SearchSuggestion from './SearchSuggestion';
import SearchTerms from './SearchTerms';
import SearchResults from './SearchResults';
import DropdownSelect from '../common/DropdownSelect';
import SearchSaveButton from './SearchSaveButton';
import Button from '../common/Button';
import { SearchContext } from '../../context/SearchContext';

// TEMP TERMS DATA TILL MOVED INTO DB
const permittedLocations = ['london', 'south east england', 'kent', 'essex'];

function SearchContainer() {
  const [search]  = useContext(SearchContext);

  // Return the various elements that make up the search view and pass the necessary props
  return (
    <div>
      {/* While props are used for common components, search functionality uses the SearchContext 
      and as a result no props are passed down */}
      {/* The dropdown selection allows the user to choose a city/location */}
      <DropdownSelect
        name="location"
        selectOptions={permittedLocations}
        value={search.location}
        onChange={search.handleDropDownSelectUpdates}
      />

      {/* The search box allows users to start typing a skill */}
      <SearchBox />

      {/* Search suggestions are rendered as the user starts typing in the search box */}
      <SearchSuggestion />

      {/* The search button triggers a fetch to the backend to return the number of jobs */}
      <Button
        buttonText="Search"
        buttonAction={search.handleSearchButtonClick}
      />

      {/* When a suggestion is clicked it is rendered as a selected search term  */}
      <SearchTerms />

      {/* As long as the user has select at least one search term and pressed search, the search
      results will be returned and displayed */}
      {search.submittedSearchTerms.length > 0 ? (
        <SearchResults />
      ) : null}

      {/* When a search is submitted by an authenticated user they will have the option to save
      their search */}
      <SearchSaveButton />

      
    </div>
  );
}

export default SearchContainer;
