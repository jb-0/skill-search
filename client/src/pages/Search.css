.image-container {
  width: 100%;
  height: 100vh;
  background-image: url('../images/home-hero.png');
  background-position: center top;
  background-repeat: no-repeat;
  background-size: cover;
  padding: var(--primary-pad) 0;
}

.image-container-mobile  {
  padding: var(--primary-pad) 0;
}

.search-container{
  width: 80%;
  margin: auto;
}

.search-container h1 {
  color: var(--secondary-color);
  text-shadow: 2px 2px var(--primary-color);
}

.search-container p {
  color: var(--secondary-color);
  text-shadow: 2px 2px var(--primary-color);
  font-size: 1.2rem;
}
