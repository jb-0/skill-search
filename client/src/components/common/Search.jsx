import React from 'react';
import './Search.css';

function Search(props) {
  return (
    <div className="search-container">
      <input
        onChange={props.handleTextBoxUpdates}
        type="text"
        id="searchInputText"
        name="searchInputText"
        autocomplete="off"
      ></input>

      {props.suggestedTerms.map((suggestedTerm) => {
        return (
          <div className="suggested-term-container">
            <p onClick={props.addSearchTerm} id={suggestedTerm}>{suggestedTerm}</p>
          </div>
        );
      })}
    </div>
  );
}

export default Search;
