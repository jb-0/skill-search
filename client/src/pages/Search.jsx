import React, { useContext } from 'react';
import './Search.css';
import Search from '../components/search/Search';
import { ViewContext } from '../context/ViewContext';

function HomePage() {
  const size = useContext(ViewContext);
  return (
    <div
      className={`image-container image-container-${size.device.toLowerCase()}`}
    >
      <div className="search-container">
        <h1>Search for your skillset</h1>
        <Search />
      </div>
    </div>
  );
}

export default HomePage;
