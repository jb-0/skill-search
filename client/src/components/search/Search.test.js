import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';

import Search from './Search';

describe('Search component', () => {
  let container = null;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('search box is successfully rendered', () => {
    act(() => {
      render(
        <Search
          suggestedTerms={[]}
          searchTerms={[]}
          searchInputText={''}
          handleTextBoxUpdates={() => {return null}}
          addSearchTerm={() => {return null}}
          removeSearchTerm={() => {return null}}
        />,
        container
      );
    });

    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('onChange callback executes when user inputs text', async () => {
    const onChange = jest.fn();

    act(() => {
      render(
        <Search
          suggestedTerms={[]}
          searchTerms={[]}
          searchInputText={''}
          handleTextBoxUpdates={onChange}
          addSearchTerm={() => {return null}}
          removeSearchTerm={() => {return null}}
        />,
        container
      );
    });

    // Imitate user typing each letter of Jav
    await userEvent.type(screen.getByRole('textbox'), 'Jav');

    expect(onChange).toHaveBeenCalledTimes(3);
  });

  it('renders correctly when passed arrays containing suggested terms and search terms', () => {
    const tree = renderer
      .create(<Search
        suggestedTerms={['node','sql']}
        searchTerms={['node','sql']}
        searchInputText={''}
        handleTextBoxUpdates={() => {return null}}
        addSearchTerm={() => {return null}}
        removeSearchTerm={() => {return null}}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when passed empty arrays', () => {
    const tree = renderer
      .create(<Search
        suggestedTerms={[]}
        searchTerms={[]}
        searchInputText={''}
        handleTextBoxUpdates={() => {return null}}
        addSearchTerm={() => {return null}}
        removeSearchTerm={() => {return null}}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });  
});
