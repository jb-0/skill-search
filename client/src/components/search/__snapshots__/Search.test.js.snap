// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component combined search components render correctly with default state 1`] = `
<form
  className="sc-eCssSg hKEeFl"
>
  <div
    className="sc-jSgupP jaIGcc"
  >
    <div
      className="dropdown-select"
    >
      <select
        id="location"
        name="location"
        onChange={[Function]}
        value="London"
      >
        <option
          value="Bristol"
        >
          Bristol
        </option>
        <option
          value="Birmingham"
        >
          Birmingham
        </option>
        <option
          value="Coventry"
        >
          Coventry
        </option>
        <option
          value="Essex"
        >
          Essex
        </option>
        <option
          value="London"
        >
          London
        </option>
        <option
          value="Kent"
        >
          Kent
        </option>
        <option
          value="Liverpool"
        >
          Liverpool
        </option>
        <option
          value="Manchester"
        >
          Manchester
        </option>
        <option
          value="Newcastle upon tyne"
        >
          Newcastle upon tyne
        </option>
      </select>
    </div>
    <div>
      <input
        autoComplete="off"
        className="sc-gsTCUz gmbZjy"
        id="searchInputText"
        name="searchInputText"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Start typing a skill.."
        type="text"
        value=""
      />
    </div>
    <button
      className="base-button undefined"
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </div>
  <div />
  <div />
</form>
`;
